Network Provider: Cloud subnets refactor

gildub opened the pull request:
Networks > Subnets > add/edit one

Replaces `$scope` with `this`
Adds `ng-cloak` and `form-change`
Removes obsolete `miqrequire`

Depends upon https://github.com/ManageIQ/manageiq-ui-classic/pull/1508

gildub commented:
@dclarizio, would you please assign it, thanks

gildub commented:
@miq-bot label networkings

gildub commented:
@miq-bot add_label networkings

gildub commented:
@miq-bot add_label networks

himdel commented:
Test failure looks valid...

gildub commented:
@himdel, all green now.

himdel commented on app/assets/javascripts/controllers/cloud_subnet/cloud_subnet_form_controller.js:
```diff
@@ -1,37 +1,50 @@
+      API.get("/api/cloud_subnets/" + cloudSubnetFormId + "?expand=resources&attributes=ext_management_system.name,cloud_tenant.name,cloud_network.name").then(function(data) {
+        vm.cloudSubnetModel.name = data.name;
+        vm.cloudSubnetModel.ems_name = data.ext_management_system.name;
+        vm.cloudSubnetModel.tenant_name = angular.isDefined(data.cloud_tenant) ? data.cloud_tenant.name : undefined;
```
FYI we're trying to remove `angular.isDefined`, this could be just `data.cloud_tenant ? data.cloud_tenant.name : undefined`.

gildub replied:
Oh, that was actually recommended recently, there are cases where the variable ends being undefined and would otherwise breaks.

himdel replied:
Well, if you *need* to check for definedness, use `data.cloud_tenant !== undefined`. 

But since you just return undefined when cloud tenant is undefined, you don't need to check that.

This could also be written as `data.cloud_tenant && data.cloud_tenant.name`.

himdel commented:
Not sure the API is quite ready for this...

I have 4 Network managers, choosing 2 of them gives me 0 tentants (that's probably OK), another one is fine, has tenants, and another one gives me...

```
GET http://localhost:3000/api/providers/10000000000019/cloud_tenants?expand=resources&attributes=id,name

{"error":{"kind":"internal_server_error","message":"undefined method `all' for nil:NilClass","klass":"NoMethodError"}}
```

After that, chosing the one with tenants and chosing a tenant and filling out a few field by random, I hit Save. .. and all I get is an error screen with `
Invalid input [cloud_subnet/create]`. No further info.

So.. I'm guessing there's some validation missing?

himdel commented:
Oh.. saving is not done via the API, it's still using `miqAjaxButton`..

so .. the `Invalid input` bit dies with this...

```
F, [2017-06-28T12:30:08.856331 #16556] FATAL -- : Error caught: [RuntimeError] Invalid input
/home/himdel/manageiq-ui-classic/app/controllers/mixins/checked_id_mixin.rb:72:in `find_record_with_rbac'
/home/himdel/manageiq-ui-classic/app/controllers/cloud_subnet_controller.rb:229:in `new_form_params'
/home/himdel/manageiq-ui-classic/app/controllers/cloud_subnet_controller.rb:65:in `create'
```

himdel commented on app/controllers/cloud_subnet_controller.rb:
```diff
@@ -261,7 +225,7 @@ def new_form_params
     end
-    options[:ip_version] = params[:network_protocol]
+    options[:ip_version] = params[:network_protocol] =~ /4/ ? 4 : 6
     options[:cloud_tenant] = find_record_with_rbac(CloudTenant, params[:cloud_tenant_id]) if params[:cloud_tenant_id]
```
.. probably caused by missing `from_cid` here.

gildub replied:
Gotcha, thanks.

gildub commented:
@himdel, this PR depends upon  #1508 which is taking care of the API get parts. Saving cannot be done by API because changes of updating model through raw commands calls wrapped in task queues. Let's wait for PR#1508 to be merged.

gildub commented:
@himdel, could you please re-review? Thanks

himdel commented:
@gildub Still missing that `from_cid`.

gildub commented:
@himdel, I've rebased and re-tested again, the creation works fine, I'm not getting any issues or any missing `from_cid`. Could you please double check?

himdel commented:
@gildub Set your database to REGION other than zero and you will, sorry :).

And yes, you *need* to use a REGION different than zero, otherwise you'll be missing from_cid from everywhere, since it is not needed for region 0.

---

Also, none of the other problems I mentioned are fixed, I'm still getting that `undefined method 'all' for nil:NilClass` when I select a different network manager, and there is no validation on the fields.

*And* the tests are failing.

Why am I even looking at this?

gildub commented:
@himdel, the problem(s) your mentioned are fixed because I couldn't reproduce it (or them) and because of lack of context/description. Also the tests are failing because an unrelated issue, and this is actually happening *way* too often maybe because we have cross repos dependencies. 

Now that said:
* I'm going to check the support feature mixin because that's supposed to take care of non supported provider
* What's the story with the database REGION? I never had to play with this, do you have any reference or documentation link?

Cheers,
Gilles

himdel commented:
> Also the tests are failing because an unrelated issue, and this is actually happening way too often maybe because we have cross repos dependencies.

Ah, thanks for restarting :). Yes, I feel your pain, it's *too* often :).


> I'm going to check the support feature mixin because that's supposed to take care of non supported provider

Thanks.. if it helps, that problematic provider...

Both providers I can see there are  `ManageIQ::Providers::Openstack::NetworkManager`. The failing one is called "OSP Director Network Manager" the working one "OpenStack Network Manager" but that probably doesn't help. 

It looks like the important part is that for the failing one, `manager.cloud_tenants` is `nil`, whereas for the other one, it returns a collection of 5 items.

I've created an issue, looks like an API bug.. https://github.com/ManageIQ/manageiq/issues/15569


> What's the story with the database REGION? I never had to play with this, do you have any reference or documentation link?

The only documentation I could find is http://manageiq.org/docs/guides/architecture/enterprise#region--enterprise . But a simple explanation is that region is a numerical prefix for all the ID values in the system.

When creating a fresh database instead of importing an existing one, if a `REGION` file exists in the `manageiq/` dir, and contains a number, the database will be created with that region setting, otherwise it will default to 0.

In region 0, `from_cid` and `to_cid` do nothing, and all objects will have purely numerical ids (like, `14`).

In regions other than 0, any id will be usually shortened in the form of `16r14` for an equivalent id in region 16. (The expanded form being `16000000000014`.) And `from_cid` is needed to convert that `16r14` back to `16000000000014`, while `to_cid` does the opposite transformation.


> this is a re-factor!

Understood, but why not make things better than refactoring? :) Or, in this case, at least useable :).

And as long as I can fill out all the fields, sucessfuly click submit, and only then get an error, the form is not quite user friendly. And when that error is a generic exception with `Invalid input`, not a flash message, and not mentioning a specific field that's wrong.. I think we can both agree that this is something that needs to be fixed, so why not now.

If you want to split this into a purely-controllerAs-refactor PR, and a fix-the-bugs PR, feel free to :).

gildub commented:
@himdel,

Sorry I missed the support feature which was present for the delete case but for the create/update cases, I just added them. 
Thanks for creating ManageIQ/manageiq#15569 .
No worries regarding the REGION, I've added the missing `from_cid`.

himdel commented:
Test failure looks familiar, restarting travis :)

himdel commented on app/assets/javascripts/controllers/cloud_subnet/cloud_subnet_form_controller.js:
```diff
@@ -1,37 +1,50 @@
+        vm.afterGet = true;
+        vm.modelCopy = angular.copy( vm.cloudSubnetModel );
+      }).catch(miqService.handleFailure);
       miqService.sparkleOff();
```
This does not belong here, should be inside that `.then` function body.. Otherwise, there is no spinner.

gildub replied:
Done

himdel commented:
controllerAs check:

   * [x] uses `as vm`
   * [ ] `ManageIQ.angular.scope` points to `vm`
   * [x] does not inject `$scope` unless needed for a `$watch` or events, or `angularForm`
   * [x] using a common button partial
   * [x] uses `miq-form`
   * [x] uses `form-changed`, no `checkchange`
   * [ ] no separate edit & new, use the same partial
   * [ ] specs
   * [ ] no copying values one by one - use `Object.assign`
   * [x] no `miqAjaxButton(url, true)` - always submit the actual model, don't rely on magic form serialization
   * [x] no extra `$setPristine` etc. where it doesn't make sense (form submit, form reset..)
   * [x] no `if (condition) return true; else return false;`
   * [x] no `miqrequired`, use `required` or `ng-required`

himdel commented:
Hey, added the checklist here as well, looks like you just removed the `manageiq.angular.scope` line, so please return it back, pointing to `vm`.

As for merging new&edit .. if this doesn't make sense to do in this PR, let me know, I can merge as is (except the scope thing).

The same goes for object.assign - would be nice to solve this here, but if there is a reason not to...

gildub commented:
@himdel,

I added back the missing `ManageIQ.angular.scope`, fix the Spinner and H3 alignment.
Yes I would rather focus on the refactor only part here.

Thanks

himdel commented:
Thanks :) Merged.. but please do look into adding some validations here.

gildub commented:
@himdel, thanks, will do.

gildub commented:
add_label fine/yes

gildub commented:
add_label fine/yes

gildub commented:
@miq-bot add_label fine/yes bug

gildub commented:
@miq-bot add_label fine/yes, bug

gildub commented:
The following PRs are taking care of the backport for Fine branch:
https://github.com/ManageIQ/manageiq/pull/16150
https://github.com/ManageIQ/manageiq-ui-classic/pull/2320

Removing `fine/yes` and 'bug' labels since the later patches are addressing the issue.

gildub commented:
@miq-bot remove label fine/yes, bug

gildub commented:
@miq-bot remove_label fine/yes, bug

simaishi commented:
Backported to Fine via https://github.com/ManageIQ/manageiq-ui-classic/pull/2320
